code/modules/xenoarchaeology/traits/xenoartifact_majors.dm

```diff
@@ -0,0 +1,589 @@
+		victims = list()
+		return
+	var/mob/living/caster = victim_a
+	var/mob/living/victim = victim_b
+	if(HAS_TRAIT(caster, TRAIT_MINDSWAPPED)) //doesn't really matter which we check
+		REMOVE_TRAIT(caster, TRAIT_MINDSWAPPED, type)
+		REMOVE_TRAIT(victim, TRAIT_MINDSWAPPED, type)
+	else
+		ADD_TRAIT(caster, TRAIT_MINDSWAPPED, type)
+		ADD_TRAIT(victim, TRAIT_MINDSWAPPED, type)
+
+	var/mob/dead/observer/ghost_v = victim.ghostize(0)
+	var/mob/dead/observer/ghost_c = caster.ghostize(0)
+	ghost_v?.mind.transfer_to(caster)
+	ghost_c?.mind.transfer_to(victim)
+	if(ghost_v?.key)
+		caster?.key = ghost_v?.key
+	if(ghost_c?.key)
+		victim?.key = ghost_c?.key
+	qdel(ghost_v)
+	qdel(ghost_c)
+
+	caster.Unconscious(6 SECONDS)
+	victim.Unconscious(6 SECONDS)
+	reverse_victims += list(list(caster, victim))
+
+/datum/xenoartifact_trait/major/mirrored/proc/undo_swap()
+	for(var/list/L as() in reverse_victims)
+		if(L.len > 1)
+			//convert to atoms to check qdel
+			var/mob/living/a = L[1]
+			var/mob/living/b = L[2]
+			if(!QDELETED(a) && !QDELETED(b))
+				swap(L[1], L[2])
+	reverse_victims = list()
+
+
+///============
+/// EMP, produces an empulse
+///============
+/datum/xenoartifact_trait/major/emp
+	label_name = "EMP"
+	label_desc = "EMP: The shape of the Artifact doesn't resemble anything particularly interesting. Technology around the Artifact seems to malfunction."
+	flags = URANIUM_TRAIT
+	weight = 25 //annoying trait
+
+/datum/xenoartifact_trait/major/emp/activate(obj/item/xenoartifact/X)
+	playsound(get_turf(X), 'sound/magic/disable_tech.ogg', 50, TRUE)
+	empulse(get_turf(X.loc), max(1, X.charge*0.03), max(1, X.charge*0.07, 1)) //This might be too big
+
+///============
+/// Invisible, makes the target invisible for a short time
+///============
+/datum/xenoartifact_trait/major/invisible //One step closer to the one ring
+	label_name = "Transparent"
+	label_desc = "Transparent: The shape of the Artifact is difficult to percieve. You feel the need to call it, precious..."
+	weight = 25
+	var/list/victims = list()
+
+/datum/xenoartifact_trait/major/invisible/on_item(obj/item/xenoartifact/X, atom/user, atom/item)
+	if(istype(item, /obj/item/laser_pointer))
+		var/obj/item/laser_pointer/L = item
+		if(L.energy)
+			to_chat(user, "<span class='info'>The [item.name]'s light passes through the structure.</span>")
+			return TRUE
+	..()
+
+/datum/xenoartifact_trait/major/invisible/activate(obj/item/xenoartifact/X, mob/living/target)
+	if(isliving(target))
+		victims += target
+		hide(target)
+		addtimer(CALLBACK(src, .proc/reveal, target), ((X.charge*0.4) SECONDS))
+		X.cooldownmod = ((X.charge*0.4)+1) SECONDS
+
+/datum/xenoartifact_trait/major/invisible/proc/hide(mob/living/target)
+	ADD_TRAIT(target, TRAIT_PACIFISM, type)
+	animate(target, ,alpha = 0, time = 5)
+
+/datum/xenoartifact_trait/major/invisible/proc/reveal(mob/living/target)
+	if(target)
+		REMOVE_TRAIT(target, TRAIT_PACIFISM, type)
+		animate(target, ,alpha = 255, time = 5)
+		target = null
```

Comment:
Setting target to null is pointless since its a local variable that immediately goes out of scope the line after